docs/book/src/cronjob-tutorial/testdata/project/api/v1/cronjob_webhook.go

```diff
@@ -0,0 +1,182 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// +kubebuilder:docs-gen:collapse=Apache License
+
+/*
+We'll start out with some imports.  You'll see below that we'll need a few more imports
+than those scaffolded for us.  We'll talk about each one when we use it.
+*/
+package v1
+
+import (
+	"github.com/robfig/cron"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apimachineryutilvalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+/*
+Next, we'll setup logger for the webhooks.
+*/
+
+var log = logf.Log.WithName("cronjob-resource")
+
+/*
+Notice that we use kubebuilder markers to generate some webhook manifests.
+The first and second markers are responsible for generating a mutating webhook manifest and a validating
+webhook manifest respectively.
+
+The meaning of each marker can be found [here](../TODO.md).
+*/
+
+// +kubebuilder:webhook:path=/mutate-batch-tutorial-kubebuilder-io-v1-cronjob,mutating=true,failurePolicy=fail,groups=batch.tutorial.kubebuilder.io,resources=cronjobs,verbs=create;update,versions=v1,name=mcronjob.kb.io
+// +kubebuilder:webhook:path=/validate-batch-tutorial-kubebuilder-io-v1-cronjob,mutating=false,failurePolicy=fail,groups=batch.tutorial.kubebuilder.io,resources=cronjobs,verbs=create;update,versions=v1,name=vcronjob.kb.io
+
+/*
+We implement the `webhook.Defaulter` interface for our CronJob CRD.
+`webhook.Defaulter` only has one method: `Default()`.
+If `webhook.Defaulter` is implemented, a webhook server will be autowired to
+serve as the mutating webhook for CronJobs.
+*/
+
+var _ webhook.Defaulter = &CronJob{}
+
+// Default implements webhookutil.defaulter so a webhook will be registered for the type
+func (c *CronJob) Default() {
+	log.Info("defaulting cronjob", "namespace", c.Namespace, "name", c.Name)
+
+	if c.Spec.ConcurrencyPolicy == "" {
+		c.Spec.ConcurrencyPolicy = AllowConcurrent
+	}
+	if c.Spec.Suspend == nil {
+		c.Spec.Suspend = new(bool)
+	}
+	if c.Spec.SuccessfulJobsHistoryLimit == nil {
+		c.Spec.SuccessfulJobsHistoryLimit = new(int32)
+		*c.Spec.SuccessfulJobsHistoryLimit = 3
+	}
+	if c.Spec.FailedJobsHistoryLimit == nil {
+		c.Spec.FailedJobsHistoryLimit = new(int32)
+		*c.Spec.FailedJobsHistoryLimit = 1
+	}
+}
+
+/*
+We implement the `webhook.Validator` interface for our CronJob CRD.
+`webhook.Validator` interface has two methods: `ValidateCreate()` and `ValidateUpdate()`.
+If `webhook.Validator` is implemented, a webhook server will be autowired to
+serve as the validating webhook for CronJobs.
```

Comment:
Talk a bit about what both do.  For example:

> The ValidateCreate method is expected to validate that its receiver is a valid object upon creation.
> If not, it can return an error that's used as a validation message.
> We separate out ValidateCreate from ValidateUpdate to allow behavior like making certain fields immutable, so that they can only be set on creation.
> Here, however, we just use the same shared validation.